File: algorithms/search/CutTheTree.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

// Solution to https://www.hackerrank.com/challenges/cut-the-tree

public class Solution {

    public static class Node {
        public int value;
        public List<Node> neighbours;
        public Node parent;
        public int sum;

        public Node(int val) {
            value = val;
            neighbours = new LinkedList<Node>();
            parent = null;
        }
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int numNodes = in.nextInt();
        Node[] arr = new Node[numNodes+1];
        int total = 0;
        for(int i = 1; i <= numNodes; i++){
            int val = in.nextInt();
            arr[i] = new Node(val);
            total += val;
        }
        for(int i = 0; i < numNodes - 1; i++){
            int vertex1 = in.nextInt();
            int vertex2 = in.nextInt();
            arr[vertex1].neighbours.add(arr[vertex2]);
            arr[vertex2].neighbours.add(arr[vertex1]);
        }
        setSum(arr[1]);

        int min = Integer.MAX_VALUE;
        for(int i = 1; i <= numNodes; i++){
            for(Node child : arr[i].neighbours){
                int diff = Math.abs(total - child.sum - child.sum);
                if(diff < min){
                    min = diff;
                }
            }
        }
        System.out.println(min);
    }

    public static int setSum(Node parent){
        int sum = parent.value;
        for(Node child : parent.neighbours){
            if(child != parent.parent){
                child.parent = parent;
                sum += setSum(child);
            }
        }
        parent.sum = sum;
        return sum;
    }
}

File: algorithms/search/cut_the_tree.py
import sys

class Node(object):
    def __init__(self, value, children=[]):
        self.value = value
        self.children = children
        self.total = 0

    def add(self, child):
        self.children.append(child)

    def setSum(self, total):
        self.total = total

n = int(input())
a = [-1]
total = 0
for x in input().split():
    total = total + int(x)
    a.append(Node(int(x), []))

for _ in range(n-1):
    u, v = [int(x) for x in input().split()]
    a[u].add(v)

def setSum(node):
    ans = a[node].value
    for child in a[node].children:
        ans = ans + setSum(child)
    a[node].setSum(ans)
    return ans

setSum(1)

trough = sys.maxsize
for i in range(1, n+1):
    for child in a[i].children:
        if total - a[child].total < trough:
            trough = abs(total - a[child].total - a[child].total)
print(trough)
